### PR TITLE
Preparation for multi-endpoint copier optimization

### PR DESCRIPTION
**Pull request by ranj063**
No change in functionality, just preparatory patches to remove the creation of DAI devices and endpoint buffers for the multi-endpoint DAI copier.

### CODE REVIEW

File: src/audio/copier/copier.c [hunk 1]
```diff
@@ -76,6 +76,49 @@ static uint32_t bitmask_to_nibble_channel_map(uint8_t bitmask)
...
+
+			channels = popcount(chan_bitmask);
+			if (channels < 1 || channels > SOF_IPC_MAX_CHANNELS) {
+				comp_err(parent_dev, "Invalid channels mask: 0x%x", chan_bitmask);
+				ret = -EINVAL;
+				goto out;
+			}
+			audio_stream_set_channels(&buffer_c->stream, channels);
+			chan_map = bitmask_to_nibble_channel_map(chan_bitmask);
+			for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
```
Comment (btian1) [inline src/audio/copier/copier.c:109]: seems line 109-110 duplicate with line 211 and line 212, please check.

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by kv2019i — APPROVED** [review]

**Comment by kv2019i** [comment]
Doh, I lost your approvals @RanderWang @lyakh , can you approve again?

**Comment by ranj063** [comment]
> looks good, is multi-endpoint tested by CI?

@lyakh not by our CI but QB has tests for it. I have requested @bardliao to look into adding a device for testing this in our CI

File: src/audio/copier/copier.c [hunk 2]
```diff
@@ -1676,68 +1676,64 @@ static int copier_params(struct comp_dev *dev, struct sof_ipc_stream_params *par
...
-					audio_stream_fmt_conversion(out_fmt->depth,
-								    out_fmt->valid_bit_depth,
-								    &out_bits, &out_valid_bits,
-								    out_fmt->s_type);
-
-					if (in_bits != in_valid_bits || out_bits != out_valid_bits)
-						cd->dd[0]->process =
-							cd->converter[IPC4_COPIER_GATEWAY_PIN];
-				} else {
-					ret = cd->endpoint[i]->drv->ops.params(cd->endpoint[i],
```
Comment (btian1) [inline src/audio/copier/copier.c:1733]: anything different with 1733（endpoint_num >1） and multi_endpoint_buffer true?  if no difference, 1733 case was removed?

**Review by btian1 — COMMENTED** [review]
ok, you want to merge preparation PR first, then use #7606 perform optimization?

File: src/audio/copier/copier.c (hunk 2 above)
Comment (ranj063) [reply src/audio/copier/copier.c:1733]: I dont understand what you're trying to say here @btian1 . ON line 1733 it says default, what case are yout alking about?

Comment (btian1) [reply src/audio/copier/copier.c:1733]: 1733 is not default, it is DAI type with endpoint_num > 1, in your changed code, there is no this case, only have multi_endpoint buffer == true(endpoint_num > 1) case(1726-1731), seems you combined, or you think right 1726 - 1731  already can cover left 1733?

Comment (ranj063) [reply src/audio/copier/copier.c:1733]: multi_endpont_buffer can exist only when endpoint_num > 1.

Comment (btian1) [reply src/audio/copier/copier.c:1733]: so line 1733 never have a chance to go here, we should remove it in our last PR, now we removed it here.

**Review by btian1 — APPROVED** [review]

**Comment by kv2019i** [comment]
One PM fail in https://sof-ci.01.org/sofpr/PR7567/build7620/devicetest/index.html
Multiple-pause-resume fail in https://sof-ci.01.org/sofpr/PR7567/build7621/devicetest/index.html -- looks like #7191 but the panic fingerprint is different.

**Pull request merged**